From dfc06954bccf87bb4540146cd9ea7926e1e7c348 Mon Sep 17 00:00:00 2001
From: usama_cowlar <usama@cowlar.com>
Date: Thu, 10 Dec 2020 13:35:38 +0500
Subject: [PATCH] custom mqtt publish message retransmit timeout set

---
 include/mqtt_client.h |  2 ++
 mqtt_client.c         | 20 ++++++++++++++++++--
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/include/mqtt_client.h b/include/mqtt_client.h
index 5d019fa..f488b6f 100644
--- a/include/mqtt_client.h
+++ b/include/mqtt_client.h
@@ -203,6 +203,7 @@ typedef struct {
     void *ds_data;                          /*!< carrier of handle for digital signature parameters */
     int network_timeout_ms;                 /*!< Abort network operation if it is not completed after this value, in milliseconds (defaults to 10s) */
     bool disable_keepalive;                 /*!< Set disable_keepalive=true to turn off keep-alive mechanism, false by default (keepalive is active by default). Note: setting the config value `keepalive` to `0` doesn't disable keepalive feature, but uses a default keepalive period */
+    int message_retransmit_timeout;
 } esp_mqtt_client_config_t;
 
 /**
@@ -347,6 +348,7 @@ int esp_mqtt_client_publish(esp_mqtt_client_handle_t client, const char *topic,
  * @return message_id if queued successfully, -1 otherwise
  */
 int esp_mqtt_client_enqueue(esp_mqtt_client_handle_t client, const char *topic, const char *data, int len, int qos, int retain, bool store);
+void esp_mqtt_client_update_retransmit_timeout(esp_mqtt_client_handle_t client, int timeout_ms);
 
 /**
  * @brief Destroys the client handle
diff --git a/mqtt_client.c b/mqtt_client.c
index ef2201d..1491dc3 100644
--- a/mqtt_client.c
+++ b/mqtt_client.c
@@ -83,6 +83,7 @@ typedef struct {
     bool skip_cert_common_name_check;
     bool use_secure_element;
     void *ds_data;
+    int message_retransmit_timeout;
 } mqtt_config_storage_t;
 
 typedef enum {
@@ -330,6 +331,15 @@ esp_err_t esp_mqtt_set_config(esp_mqtt_client_handle_t client, const esp_mqtt_cl
         cfg->task_prio = config->task_prio;
     }
 
+    if (config->message_retransmit_timeout > 0) {
+        cfg->message_retransmit_timeout = config->message_retransmit_timeout;
+    }
+    else
+    {
+       cfg->message_retransmit_timeout = 1000;
+    }
+    
+
     if (cfg->task_prio <= 0) {
         cfg->task_prio = MQTT_TASK_PRIORITY;
     }
@@ -1396,10 +1406,10 @@ static void esp_mqtt_task(void *pv)
                     outbox_set_pending(client->outbox, client->mqtt_state.pending_msg_id, TRANSMITTED);
                 }
                 // resend other "transmitted" messages after 1s
-            } else if (platform_tick_get_ms() - last_retransmit > 1000) {
+            } else if (platform_tick_get_ms() - last_retransmit > client->config->message_retransmit_timeout) {
                 last_retransmit = platform_tick_get_ms();
                 item = outbox_dequeue(client->outbox, TRANSMITTED, &msg_tick);
-                if (item && (last_retransmit - msg_tick > 1000))  {
+                if (item && (last_retransmit - msg_tick > client->config->message_retransmit_timeout))  {
                     mqtt_resend_queued(client, item);
                 }
             }
@@ -1833,3 +1843,9 @@ int esp_mqtt_client_get_outbox_size(esp_mqtt_client_handle_t client)
 
     return outbox_size;
 }
+
+void esp_mqtt_client_update_retransmit_timeout(esp_mqtt_client_handle_t client, int timeout_ms)
+{
+    if (client)
+        client->config->message_retransmit_timeout = timeout_ms;
+}
-- 
2.18.0.windows.1

